File: src/disk_tree/staged.py
"""The staged-delete engine (spec ``specs/staged-delete.md``, checkpoint 1).

Stage URIs into a shared open :class:`~disk_tree.sqla.Plan`, then dispatch the
plan — a :class:`~disk_tree.sqla.DeletionRun` that deletes each item (real) or
reports it (dry), recording a per-URI :class:`~disk_tree.sqla.DeletionBand`.

Deployment-agnostic: ``delete_fn(uri)`` (what actually removes the object — the
existing ``backend_for(uri).delete``) and ``size_fn(uri) -> (bytes, objects)``
are injected, so the engine has no backend or DB-init coupling and unit-tests
against a plain session with fakes.
"""
from __future__ import annotations

import secrets
from datetime import datetime
from typing import Callable, Iterable

from sqlalchemy import select
from sqlalchemy.orm import Session

from .sqla import DeletionBand, DeletionRun, Plan, PlanItem

STAGED = "Staged"

SizeFn = Callable[[str], "tuple[int, int]"]
DeleteFn = Callable[[str], None]


def _now() -> datetime:
    return datetime.now().astimezone()


def _canonical(uri: str) -> str:
    from disk_tree.backends import canonical

    return canonical(uri)


def open_plan(session: Session, who: str, name: str = STAGED) -> Plan:
    """The shared open plan named ``name``, created (empty) if none is open."""
    plan = session.scalars(
        select(Plan).where(Plan.name == name, Plan.state == "open").order_by(Plan.id.desc())
    ).first()
    if plan is None:
        plan = Plan(name=name, created_by=who, created_ts=_now())
        session.add(plan)
        session.flush()
    return plan


def stage(session: Session, uris: Iterable[str], who: str, note: str | None = None) -> tuple[Plan, list[str]]:
    """Add ``uris`` to the shared open plan (idempotent per URI). Returns the
    plan and the URIs newly staged (already-staged ones are skipped)."""
    plan = open_plan(session, who)
    have = set(session.scalars(select(PlanItem.uri).where(PlanItem.plan_id == plan.id)))
    added: list[str] = []
    for raw in uris:
        uri = _canonical(raw)
        if uri in have:
            continue
        session.add(PlanItem(plan_id=plan.id, uri=uri, added_by=who, added_ts=_now(), note=note))
        have.add(uri)
        added.append(uri)
    session.flush()
    return plan, added


def unstage(session: Session, uris: Iterable[str]) -> int:
    """Remove ``uris`` from every open plan. Returns the number removed."""
    open_ids = set(session.scalars(select(Plan.id).where(Plan.state == "open")))
    removed = 0
    for raw in uris:
        uri = _canonical(raw)
        for item in session.scalars(select(PlanItem).where(PlanItem.uri == uri, PlanItem.plan_id.in_(open_ids))):
            session.delete(item)
            removed += 1
    session.flush()
    return removed


def plan_by_ref(session: Session, ref: str | None) -> Plan | None:
    """Resolve a plan by id (numeric) or name; ``None`` -> the open ``Staged`` plan."""
    if ref is None:
        return session.scalars(
            select(Plan).where(Plan.name == STAGED, Plan.state == "open").order_by(Plan.id.desc())
        ).first()
    if ref.isdigit():
        return session.get(Plan, int(ref))
    return session.scalars(select(Plan).where(Plan.name == ref).order_by(Plan.id.desc())).first()


def items(session: Session, plan: Plan) -> list[PlanItem]:
    return list(session.scalars(select(PlanItem).where(PlanItem.plan_id == plan.id).order_by(PlanItem.uri)))


def dispatch(
    session: Session,
    plan: Plan,
    actor: str,
    *,
    for_real: bool,
    delete_fn: DeleteFn,
    size_fn: SizeFn,
    undo_state: str = "none",
) -> DeletionRun:
    """Execute ``plan``: size each item (``size_fn``) and, when ``for_real``,
    delete it (``delete_fn``). Records a :class:`DeletionRun` + one
    :class:`DeletionBand` per item; a real dispatch closes the plan. ``dry`` (the
    default) deletes nothing and leaves the plan open."""
    started = _now()
    # a plan takes many dry runs + one real, so the second-resolution timestamp
    # alone collides; a random suffix keeps run_ids unique
    run_id = f"{plan.id}-{started.strftime('%Y%m%dT%H%M%S')}-{secrets.token_hex(4)}"
    run = DeletionRun(
        run_id=run_id, plan_id=plan.id, mode="real" if for_real else "dry",
        actor=actor, started_ts=started, undo_state=undo_state if for_real else "none",
    )
    session.add(run)
    for item in items(session, plan):
        nbytes, nobjs = size_fn(item.uri)
        deleted = 0
        if for_real:
            delete_fn(item.uri)
            deleted = 1
            run.deleted_bytes += nbytes
            run.deleted_objects += nobjs
        session.add(DeletionBand(run_id=run_id, uri=item.uri, bytes=nbytes, objects=nobjs, deleted=deleted))
    run.finished_ts = _now()
    if for_real:
        plan.state = "closed"
        plan.closed_ts = run.finished_ts
    session.flush()
    return run

File: ui/cfn/staged.ts
/**
 * The staged-delete data layer over D1 (spec `specs/staged-delete.md`, CP2) —
 * the edge mirror of the CP1 Python engine (`src/disk_tree/staged.py`).
 *
 * A shared open `plan` collects `plan_items` (trash gestures, `view` scope);
 * an admin dispatch (`admin` scope) *enqueues* a real `deletion_run` and closes
 * the plan. The edge can't reach arbitrary user buckets, so it never deletes:
 * an enqueued run (`finished_ts IS NULL`) is drained by the server-side
 * executor (CP4), which sizes each item and lands its `deletion_band`.
 */

export const STAGED = 'Staged'

export interface Plan {
  id: number
  name: string
  note: string | null
  state: 'open' | 'closed'
  created_by: string
  created_ts: number
  closed_ts: number | null
}

export interface PlanItem {
  plan_id: number
  uri: string
  note: string | null
  added_by: string
  added_ts: number
}

export interface DeletionRun {
  run_id: string
  plan_id: number
  mode: 'dry' | 'real'
  actor: string
  started_ts: number
  finished_ts: number | null
  deleted_bytes: number
  deleted_objects: number
  skipped_gone: number
  undo_state: string
  undo_deadline: number | null
}

const nowS = (): number => Math.floor(Date.now() / 1000)

/** URI as the CP1 engine canonicalizes it: no trailing slash (schemes keep
 *  their `//`, since only *trailing* slashes are stripped). */
export const canonicalUri = (raw: string): string => raw.replace(/\/+$/, '')

/** `YYYYMMDDTHHMMSS` (UTC) for a run id, mirroring the Python `strftime`. */
function stamp(ts: number): string {
  const d = new Date(ts * 1000)
  const p = (n: number, w = 2) => String(n).padStart(w, '0')
  return (
    `${d.getUTCFullYear()}${p(d.getUTCMonth() + 1)}${p(d.getUTCDate())}` +
    `T${p(d.getUTCHours())}${p(d.getUTCMinutes())}${p(d.getUTCSeconds())}`
  )
}

function hex4(): string {
  const b = new Uint8Array(4)
  crypto.getRandomValues(b)
  return Array.from(b, x => x.toString(16).padStart(2, '0')).join('')
}

/** The shared open plan named `name`, created (empty) if none is open. */
export async function openPlan(db: D1Database, who: string, name: string = STAGED): Promise<Plan> {
  const found = await db
    .prepare(`SELECT * FROM plans WHERE name = ?1 AND state = 'open' ORDER BY id DESC LIMIT 1`)
    .bind(name)
    .first<Plan>()
  if (found) return found
  const ts = nowS()
  const { meta } = await db
    .prepare(`INSERT INTO plans (name, created_by, created_ts) VALUES (?1, ?2, ?3)`)
    .bind(name, who, ts)
    .run()
  return { id: Number(meta.last_row_id), name, note: null, state: 'open', created_by: who, created_ts: ts, closed_ts: null }
}

/** Add `uris` to the shared open plan (idempotent per URI). Returns the plan
 *  and the URIs newly staged. */
export async function stage(
  db: D1Database,
  uris: string[],
  who: string,
  note: string | null = null,
): Promise<{ plan: Plan; added: string[] }> {
  const plan = await openPlan(db, who)
  const have = new Set(
    (await db.prepare(`SELECT uri FROM plan_items WHERE plan_id = ?1`).bind(plan.id).all<{ uri: string }>()).results.map(
      r => r.uri,
    ),
  )
  const ts = nowS()
  const added: string[] = []
  const inserts: D1PreparedStatement[] = []
  for (const raw of uris) {
    const uri = canonicalUri(raw)
    if (have.has(uri)) continue
    have.add(uri)
    added.push(uri)
    inserts.push(
      db
        .prepare(`INSERT INTO plan_items (plan_id, uri, note, added_by, added_ts) VALUES (?1, ?2, ?3, ?4, ?5)`)
        .bind(plan.id, uri, note, who, ts),
    )
  }
  if (inserts.length) await db.batch(inserts)
  return { plan, added }
}

/** Remove `uris` from every open plan. Returns the number removed. */
export async function unstage(db: D1Database, uris: string[]): Promise<number> {
  let removed = 0
  for (const raw of uris) {
    const { meta } = await db
      .prepare(
        `DELETE FROM plan_items WHERE uri = ?1 AND plan_id IN (SELECT id FROM plans WHERE state = 'open')`,
      )
      .bind(canonicalUri(raw))
      .run()
    removed += meta.changes ?? 0
  }
  return removed
}

/** Resolve a plan by id (numeric) or name; `null` ref -> the open `Staged` plan. */
export async function planByRef(db: D1Database, ref: string | null): Promise<Plan | null> {
  if (ref === null) {
    return db
      .prepare(`SELECT * FROM plans WHERE name = ?1 AND state = 'open' ORDER BY id DESC LIMIT 1`)
      .bind(STAGED)
      .first<Plan>()
  }
  if (/^\d+$/.test(ref)) return db.prepare(`SELECT * FROM plans WHERE id = ?1`).bind(Number(ref)).first<Plan>()
  return db.prepare(`SELECT * FROM plans WHERE name = ?1 ORDER BY id DESC LIMIT 1`).bind(ref).first<Plan>()
}

export async function planItems(db: D1Database, planId: number): Promise<PlanItem[]> {
  const { results } = await db
    .prepare(`SELECT * FROM plan_items WHERE plan_id = ?1 ORDER BY uri`)
    .bind(planId)
    .all<PlanItem>()
  return results
}

export async function listOpenPlans(db: D1Database): Promise<Plan[]> {
  const { results } = await db.prepare(`SELECT * FROM plans WHERE state = 'open' ORDER BY id`).all<Plan>()
  return results
}

export async function listRuns(db: D1Database, limit = 10): Promise<DeletionRun[]> {
  const { results } = await db
    .prepare(`SELECT * FROM deletion_runs ORDER BY started_ts DESC LIMIT ?1`)
    .bind(limit)
    .all<DeletionRun>()
  return results
}

/** Dispatch a plan from the edge: enqueue a real run (pending — the executor
 *  drains it) and close the plan so its staged set is frozen. No bands: the
 *  server-side executor produces them with real sizes. */
export async function enqueueDispatch(db: D1Database, plan: Plan, actor: string): Promise<DeletionRun> {
  const ts = nowS()
  const run_id = `${plan.id}-${stamp(ts)}-${hex4()}`
  await db.batch([
    db
      .prepare(
        `INSERT INTO deletion_runs (run_id, plan_id, mode, actor, started_ts) VALUES (?1, ?2, 'real', ?3, ?4)`,
      )
      .bind(run_id, plan.id, actor, ts),
    db.prepare(`UPDATE plans SET state = 'closed', closed_ts = ?2 WHERE id = ?1`).bind(plan.id, ts),
  ])
  return {
    run_id, plan_id: plan.id, mode: 'real', actor, started_ts: ts, finished_ts: null,
    deleted_bytes: 0, deleted_objects: 0, skipped_gone: 0, undo_state: 'none', undo_deadline: null,
  }
}
